containous/traefik
# Add frontend, backend, and timing to access log

This PR enhances the access log to include frontend and backend names and total time.  Log format is:

    127.0.0.1 - - [14/Apr/2016:16:30:45 +0000] "GET /test HTTP/1.0" 200 31 "" "ApacheBench/2.3" 334110 "frontend" "http://127.0.0.1:8081" 56.846528ms

The PR adds the last four tokens:

* reqid (a count of the total number of requests since Traefik startup)
* frontend name (derived from the backend name via a map built from the current configuration)
* backend URL (assigned by oxy)
* elapsed time

This PR is a successor to PR #215 which was closed unmerged after its author left the project. 

#### Overview of Update
1. middlewares/logger.go is substantially revised to gather and report the new information.  It creates the reqid and stores it in a new custom header, X-Traefik-Reqid that can be accessed when subsequent middleware methods call function saveBackendNameForLogger(). The reqid provides a context for storing names. Note that the HTTP request pointer cannot be used as a context key (making Gorilla context unsuitable) because oxy replaces the HTTP request with a shallow copy when assigning a backend. The logger deletes the custom header before returning to its caller.

2. middlewares/saveBackend.go runs after oxy has assigned a backend and calls saveBackendNameForLogger() to report it to the logger, which saves it in a mutex-protected map.

#### Testing
Three tests are provided with the PR

* A unit test for logger.go in middlewares/logger_test.go.   Run with `go test` from the middlewares directory.

* A functional test in examples/accessLog.  This test starts three http servers and provides a curl command to send requests to them so that the access.log update can be verified. To run it:
  1. Run script examples/accessLog/runExample.sh from the Traefik root directory.
  2. Run command `curl http://127.0.0.1:80/test{1,2,2}` to send some requests.
  3. Run command `tail log/access.log` to verify the access.log was written correctly.

* A benchmark stress test derived from the functional test above.  This uses [Apache ab](https://httpd.apache.org/docs/2.4/programs/ab.html), which must be pre-installed.   Use script examples/accessLog/runAb.sh to run it.  This script supports several options for setting the number of connections, number of requests, and total time of the test.  Use command `runAb.sh --help` to list them.

#### Response to Comments
@emilevauge [commented](https://github.com/containous/traefik/pull/215#issuecomment-199936524) on the previous PR to make some requests:

* could you squash your commits?
  _The repo has been recreated and has only two commits, with all source updates in the first._
* rather than a bash script test, I would prefer integration or unit tests of your code ;)
  _The PR now includes a unit test in addition to the script_
* I would rename routes to saveFrontend to stay homogeneous
  _routes.go is no longer part of the PR, as recent changes to server.go broke its functionality.  The PR now builds a backend-to-frontend map and derives the frontend name from the backend URL.  FYI, the server change that broke saveFrontend was to add an unnamed subrouter to the router with the path.  mux.RouteMatch() matches both, but returns only the unnamed subrouter._
* Did you make benchmarks on traefik using that (using ab, or wrk), to test high loads?
_The PR includes a benchmark script using ab.  I used it to run a one-hour test with 100 concurrent connections, which showed no memory leaks or concurrency issues._

Sorry for the delay in responding to these comments; I got pulled into another project for a time.
